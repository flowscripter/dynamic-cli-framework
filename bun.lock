{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "@flowscripter/dynamic-plugin-framework",
      "dependencies": {
        "emphasize": "latest",
        "fastest-levenshtein": "latest",
        "figlet": "latest",
        "highlight.js": "latest",
        "prettier": "latest",
        "supports-color": "latest",
        "ts-log": "latest",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "latest",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.21", "", { "dependencies": { "bun-types": "1.2.21" } }, "sha512-NiDnvEqmbfQ6dmZ3EeUO577s4P5bf4HCTXtI6trMc6f6RzirY5IrF3aIookuSpyslFzrnvv2lmEWv5HyC1X79A=="],

    "@types/hast": ["@types/hast@3.0.4", "", { "dependencies": { "@types/unist": "*" } }, "sha512-WPs+bbQw5aCj+x6laNGWLH3wviHtoCv/P3+otBhbOhJgG8qtpdAMlTCxLtsTWA7LH1Oh/bFCHsBn0TPS5m30EQ=="],

    "@types/node": ["@types/node@22.13.4", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-ywP2X0DYtX3y08eFVx5fNIw7/uIv8hYUKgXoK8oayJlLnKcRfEYCxWMVE1XagUdVtCJlZT1AU4LXEABW+L1Peg=="],

    "@types/react": ["@types/react@19.1.8", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-AwAfQ2Wa5bCx9WP8nZL2uMZWod7J7/JSplxbTmBQ5ms6QpqNYm672H0Vu9ZVKVngQ+ii4R/byguVEUZQyeg44g=="],

    "@types/unist": ["@types/unist@3.0.3", "", {}, "sha512-ko/gIFJRv177XgZsZcBwnqJN5x/Gien8qNOn0D5bQU/zAzVf9Zt3BlcUiLqhV9y4ARk0GbT3tnUiPNgnTXzc/Q=="],

    "bun-types": ["bun-types@1.2.21", "", { "dependencies": { "@types/node": "*" }, "peerDependencies": { "@types/react": "^19" } }, "sha512-sa2Tj77Ijc/NTLS0/Odjq/qngmEPZfbfnOERi0KRUYhT9R8M4VBioWVmMWE5GrYbKMc+5lVybXygLdibHaqVqw=="],

    "chalk": ["chalk@5.4.1", "", {}, "sha512-zgVZuo2WcZgfUEmsn6eO3kINexW8RAE4maiQ8QNs8CtpPCSyMiYsULR3HQYkm3w8FIA3SberyMJMSldGsW+U3w=="],

    "commander": ["commander@14.0.0", "", {}, "sha512-2uM9rYjPvyq39NwLRqaiLtWHyDC1FvryJDa2ATTVims5YAS4PupsEQsDvP14FqhFr0P49CYDugi59xaxJlTXRA=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "dequal": ["dequal@2.0.3", "", {}, "sha512-0je+qPKHEMohvfRTCEo3CrPG6cAzAYgmzKyxRiYSSDkS6eGJdyVJm7WaYA5ECaAD9wLB2T4EEeymA5aFVcYXCA=="],

    "devlop": ["devlop@1.1.0", "", { "dependencies": { "dequal": "^2.0.0" } }, "sha512-RWmIqhcFf1lRYBvNmr7qTNuyCt/7/ns2jbpp1+PalgE/rDQcBT0fioSMUpJ93irlUhC5hrg4cYqe6U+0ImW0rA=="],

    "emphasize": ["emphasize@7.0.0", "", { "dependencies": { "@types/hast": "^3.0.0", "chalk": "^5.0.0", "highlight.js": "~11.9.0", "lowlight": "~3.1.0" } }, "sha512-jdFCDyt+YetBXO12VwK4AiLsMCvkZ3IBxMVIJddB+25EwIL0VETBgpvPkJl63+JyAgaQ5Wja10qWMoXXC95JNg=="],

    "fastest-levenshtein": ["fastest-levenshtein@1.0.16", "", {}, "sha512-eRnCtTTtGZFpQCwhJiUOuxPQWRXVKYDn0b2PeHfXL6/Zi53SLAzAHfVhVWK2AryC/WH05kGfxhFIPvTF0SXQzg=="],

    "figlet": ["figlet@1.9.0", "", { "dependencies": { "commander": "^14.0.0" }, "bin": { "figlet": "bin/index.js" } }, "sha512-S+WN3ZIGYgiZUlQD0s3qlAPSOdCdYth0G7PK6DbubfixgorA7jzNl4eae2Hc/rEjQ5UYnf0YkcpCFc/IMklZhg=="],

    "highlight.js": ["highlight.js@11.11.1", "", {}, "sha512-Xwwo44whKBVCYoliBQwaPvtd/2tYFkRQtXDWj1nackaV2JPXx3L0+Jvd8/qCJ2p+ML0/XVkJ2q+Mr+UVdpJK5w=="],

    "lowlight": ["lowlight@3.1.0", "", { "dependencies": { "@types/hast": "^3.0.0", "devlop": "^1.0.0", "highlight.js": "~11.9.0" } }, "sha512-CEbNVoSikAxwDMDPjXlqlFYiZLkDJHwyGu/MfOsJnF3d7f3tds5J3z8s/l9TMXhzfsJCCJEAsD78842mwmg0PQ=="],

    "prettier": ["prettier@3.6.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-I7AIg5boAr5R0FFtJ6rCfD+LFsWHp81dolrFD8S79U9tb8Az2nGrJncnMSnys+bpQJfRUzqs9hnA81OAA3hCuQ=="],

    "supports-color": ["supports-color@10.2.2", "", {}, "sha512-SS+jx45GF1QjgEXQx4NJZV9ImqmO2NPz5FNsIHrsDjh2YsHnawpan7SNQ1o8NuhrbHZy9AZhIoCUiCeaW/C80g=="],

    "ts-log": ["ts-log@2.2.7", "", {}, "sha512-320x5Ggei84AxzlXp91QkIGSw5wgaLT6GeAH0KsqDmRZdVWW2OiSeVvElVoatk3f7nicwXlElXsoFkARiGE2yg=="],

    "typescript": ["typescript@5.9.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-CWBzXQrc/qOkhidw1OzBTQuYRbfyxDXJMVJ1XNwUHGROVmuaeiEm3OslpZ1RV96d7SKKjZKrSJu3+t/xlw3R9A=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],

    "emphasize/highlight.js": ["highlight.js@11.9.0", "", {}, "sha512-fJ7cW7fQGCYAkgv4CPfwFHrfd/cLS4Hau96JuJ+ZTOWhjnhoeN1ub1tFmALm/+lW5z4WCAuAV9bm05AP0mS6Gw=="],

    "lowlight/highlight.js": ["highlight.js@11.9.0", "", {}, "sha512-fJ7cW7fQGCYAkgv4CPfwFHrfd/cLS4Hau96JuJ+ZTOWhjnhoeN1ub1tFmALm/+lW5z4WCAuAV9bm05AP0mS6Gw=="],
  }
}
