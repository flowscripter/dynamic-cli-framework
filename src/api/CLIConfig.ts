/**
 * Interface specifying common configuration for a {@link CLI} application.
 */
export default interface CLIConfig {
  /**
   * Name of the CLI application.
   */
  readonly name: string;

  /**
   * Optional description of the CLI application.
   */
  readonly description?: string;

  /**
   * Version of the CLI application.
   */
  readonly version: string;
}
