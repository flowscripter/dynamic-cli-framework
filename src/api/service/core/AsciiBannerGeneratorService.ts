export const ASCII_BANNER_GENERATOR_SERVICE_ID =
  "@flowscripter/dynamic-cli-framework/ascii-banner-generator-service";

/**
 * Service allowing a text string to be rendered using an ASCII banner [FIGlet](http://www.figlet.org) font.
 */
export default interface AsciiBannerGeneratorService {
  /**
   * Register a new font.
   *
   * The recommended way to load the font definition is to convert the font definition to a JSON string
   * (replacing newlines with `\n`) and placing in a simple JSON file: `{ "font": "<figlet font definition>" }`.
   * This can then be imported as follows:
   *
   * `import { font as myFont } from "./myfont.json" with { type: "json" };
   *
   * @param fontName the name used to refer to the font.
   * @param fontDefinition the definition for the syntax conforming to the
   * [FIGfont format](http://www.jave.de/figlet/figfont.html#creating).
   */
  registerFont(fontName: string, fontDefinition: string): void;

  /**
   * Return the names of the currently registered fonts.
   */
  getRegisteredFonts(): ReadonlyArray<string>;

  /**
   * Generate an ASCII banner text for the message using the specified font.
   *
   * @param message the message to output.
   * @param fontName the font to use.
   */
  generate(message: string, fontName: string): Promise<string>;
}
