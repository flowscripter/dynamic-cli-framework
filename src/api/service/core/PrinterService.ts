export const PRINTER_SERVICE_ID =
  "@flowscripter/dynamic-cli-framework/printer-service";
import { WritableStream } from "node:stream/web";

/**
 * Enum of message importance level.
 */
export enum Level {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

/**
 * Enum of message icons.
 */
export enum Icon {
  // Will be displayed in green if {@link Printer.colorEnabled} is `true`.
  SUCCESS = 0,
  // Will be displayed in red if {@link Printer.colorEnabled} is `true`.
  FAILURE = 1,
  // Will be displayed in yellow if {@link Printer.colorEnabled} is `true`.
  ALERT = 2,
  // Will be displayed in blue if {@link Printer.colorEnabled} is `true`.
  INFORMATION = 3,
}

/**
 * Service allowing a {@link Command} to output user messages to `stdout` and/or `stderr`.
 *
 * Output to `stdout` is via {@link print} whilst output to `stderr` is via a filtered logging mechanism
 * using {@link debug}, {@link info}, {@link warn} and {@link error}.
 */
export default interface PrinterService {
  /**
   * Disable or enable color output for messages.
   */
  colorEnabled: boolean;

  /**
   * Enable or disable dark mode. Default is disabled i.e. `false`.
   */
  darkMode: boolean;

  /**
   * The WritableStream used for `stdout`. Can be accessed directly for output of binary data etc.
   */
  stdoutWritable: WritableStream;

  /**
   * The WritableStream used for `stderr`. Can be accessed directly for output of binary data etc.
   */
  stderrWritable: WritableStream;

  /**
   * Return the provided message so that the foreground is colored as primary content. This is the default
   * calor applied. The actual color will depend on the value of {@link darkMode}.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  primary(message: string): string;

  /**
   * Return the provided message so that the foreground is colored as secondary content.
   * The actual color will depend on the value of {@link darkMode}.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  secondary(message: string): string;

  /**
   * Return the provided message so that the foreground is colored as emphasised content.
   * The actual color will depend on the value of {@link darkMode}.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  emphasised(message: string): string;

  /**
   * Return the provided message so that the background is colored as selected content.
   * The actual color will depend on the value of {@link darkMode}.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  selected(message: string): string;

  /**
   * Return the provided message so that the text is displayed in italic.
   */
  italic(message: string): string;

  /**
   * Return the provided message so that the foreground is yellow.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  yellow(message: string): string;

  /**
   * Return the provided message so that the foreground is orange.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  orange(message: string): string;

  /**
   * Return the provided message so that the foreground is red.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  red(message: string): string;

  /**
   * Return the provided message so that the foreground is magenta.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  magenta(message: string): string;

  /**
   * Return the provided message so that the foreground is violet.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  violet(message: string): string;

  /**
   * Return the provided message so that the foreground is blue.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  blue(message: string): string;

  /**
   * Return the provided message so that the foreground is cyan.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  cyan(message: string): string;

  /**
   * Return the provided message so that the foreground is green.
   * Has no effect if {@link colorEnabled} is `false`.
   */
  green(message: string): string;

  /**
   * Return the provided message so that the foreground is the specified color.
   * Has no effect if {@link colorEnabled} is `false`.
   *
   * @param message the message to color.
   * @param hexFormattedColor the color to use. This should be a valid hex formatted string e.g. "0xrrggbb".
   */
  color(message: string, hexFormattedColor: string): string;

  /**
   * Print a message on `stdout`.
   * Will be displayed as primary content if {@link colorEnabled} is `true`.
   *
   * @param message the message to output.
   * @param icon optional icon to display with the message.
   */
  print(message: string, icon?: Icon): Promise<void>;

  /**
   * Print a {@link DEBUG} level message on `stderr`.
   * Will be displayed as secondary content if {@link colorEnabled} is `true`.
   *
   * @param message the message to output.
   * @param icon optional icon to display with the message.
   */
  debug(message: string, icon?: Icon): Promise<void>;

  /**
   * Print an {@link INFO} level message on `stderr`.
   * Will be displayed as primary content if {@link colorEnabled} is `true`.
   *
   * @param message the message to output.
   * @param icon optional icon to display with the message.
   */
  info(message: string, icon?: Icon): Promise<void>;

  /**
   * Print a {@link WARN} level message on `stderr`.
   * Will be displayed as yellow content if {@link colorEnabled} is `true`.
   *
   * @param message the message to output.
   * @param icon optional icon to display with the message.
   */
  warn(message: string, icon?: Icon): Promise<void>;

  /**
   * Print an {@link ERROR} level message on `stderr`.
   * Will be displayed as red content if {@link colorEnabled} is `true`.
   *
   * @param message the message to output.
   * @param icon optional icon to display with the message.
   */
  error(message: string, icon?: Icon): Promise<void>;

  /**
   * Set the output threshold {@link Level} for `stderr`.
   *
   * Default level is {@link INFO}.
   *
   * @param level any message below this level will be filtered from output,
   */
  setLevel(level: Level): void;

  /**
   * Get the output threshold {@link Level} for `stderr`.
   */
  getLevel(): Level;

  /**
   * Display the spinner on `stderr`.
   *
   * The spinner will be displayed as emphasised content and the message will
   * be displayed as primary content if {@link colorEnabled} is `true`.
   *
   * NOTE: The spinner and message will be displayed at {@link INFO} level.
   *
   * NOTE: If the spinner is already displayed the message will be updated to that specified.
   *
   * NOTE: If any progress bars are currently displayed they will be hidden.
   *
   * @param message the message to output after the spinner.
   */
  showSpinner(message: string): Promise<void>;

  /**
   * Hide the spinner.
   *
   * NOTE: Showing a progress bar will also hide the spinner.
   */
  hideSpinner(): Promise<void>;

  /**
   * Display a progress bar on `stderr`.
   *
   * The progress will be displayed in green if {@link colorEnabled} is `true`.
   *
   * NOTE: The progress bar and message will be displayed at {@link INFO} level.
   *
   * NOTE: If the spinner is currently displayed it will be hidden.
   *
   * @param units the units to display for progress indication e.g. 'MB' or 'Kb'.
   * @param message an optional message for the progress bar.
   * @param total the total value which equates to 100% complete, defaults to `100`.
   * @param current the current value which is a portion of the total value, defaults to `0`.
   *
   * @return a handle to use when invoking {@link updateProgressBar}.
   */
  showProgressBar(
    units: string,
    message?: string,
    total?: number,
    current?: number,
  ): Promise<number>;

  /**
   * Hides a specified progress bar.
   *
   * NOTE: Showing the spinner will also hide ALL progress bars.
   *
   * @param handle the handle referring to the progress bar to be hidden.
   */
  hideProgressBar(handle: number): Promise<void>;

  /**
   * Hides all progress bars.
   *
   * NOTE: Showing the spinner will also hide ALL progress bars.
   */
  hideAllProgressBars(): Promise<void>;

  /**
   * Update a specific progress bar.
   *
   * @param handle the handle referring to the progress bar to be updated.
   * @param current the current value to set on the progress bar.
   * @param message an optional message to set on the progress bar, if not specified the initially specified message will be displayed.
   */
  updateProgressBar(handle: number, current: number, message?: string): void;
}
