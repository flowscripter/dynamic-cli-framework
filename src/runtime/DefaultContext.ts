import type Context from "../api/Context.ts";
import type CLIConfig from "../api/CLIConfig.ts";

/**
 * Default implementation of a {@link Context}.
 */
export default class DefaultContext implements Context {
  readonly #serviceInstancesById: Map<string, unknown> = new Map();

  public constructor(readonly cliConfig: CLIConfig) {
  }

  public addServiceInstance(id: string, serviceInstance: unknown) {
    if (this.#serviceInstancesById.has(id)) {
      throw new Error(
        `Service ID: ${id} duplicates the ID of an existing service`,
      );
    }
    this.#serviceInstancesById.set(id, serviceInstance);
  }

  public getServiceById(id: string): unknown {
    if (!this.#serviceInstancesById.has(id)) {
      throw new Error(`Service with ID: ${id} does not exist`);
    }
    return this.#serviceInstancesById.get(id);
  }

  public doesServiceExist(id: string): boolean {
    return this.#serviceInstancesById.has(id);
  }
}
