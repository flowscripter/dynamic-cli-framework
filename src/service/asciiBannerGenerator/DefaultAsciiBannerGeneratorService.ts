import type AsciiBannerGeneratorService from "../../api/service/core/AsciiBannerGeneratorService.ts";
import figlet from "figlet";
// FIGlet font is converted to a JSON string and embedded in a simple JSON file: `{ "font": "<figlet font definition>" }`
import standardFont from "./standard.flf.json" with { type: "json" };

/**
 * Default implementation of {@link AsciiBannerGeneratorService} which has a font definition
 * for the FIGlet "standard" font already registered.
 */
export default class DefaultAsciiBannerGeneratorService
  implements AsciiBannerGeneratorService {
  #registeredFontNames: Array<string> = [];

  constructor() {
    this.registerFont("standard", standardFont.font);
  }

  getRegisteredFonts(): ReadonlyArray<string> {
    return this.#registeredFontNames;
  }

  registerFont(fontName: string, fontDefinition: string): void {
    const name = fontName.toLowerCase();
    if (this.getRegisteredFonts().includes(name)) {
      throw new Error(`Font name already registered: ${name}`);
    }
    this.#registeredFontNames.push(name);
    figlet.parseFont(fontName, fontDefinition);
  }

  async generate(message: string, fontName: string): Promise<string> {
    const name = fontName.toLowerCase();
    if (!this.getRegisteredFonts().includes(name)) {
      throw new Error(`Font name is not registered: ${name}`);
    }

    const bannerText = await figlet.text(message, { font: name });
    const lines = bannerText.split("\n");

    // Remove any whitespace lines from the end
    while (
      (lines.length > 0) && (lines[lines.length - 1].trim().length === 0)
    ) {
      lines.pop();
    }

    return "\n" + lines.join("\n") + "\n";
  }
}
