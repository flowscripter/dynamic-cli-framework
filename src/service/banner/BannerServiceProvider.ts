import type Context from "../../api/Context.ts";
import type {
  ServiceInfo,
  ServiceProvider,
} from "../../api/service/ServiceProvider.ts";
import NoBannerCommand from "./command/NoBannerCommand.ts";
import type PrinterService from "../../api/service/core/PrinterService.ts";
import { PRINTER_SERVICE_ID } from "../../api/service/core/PrinterService.ts";
import type ConfigurationServiceProvider from "../configuration/ConfigurationServiceProvider.ts";
import type AsciiBannerGeneratorService from "../../api/service/core/AsciiBannerGeneratorService.ts";
import {
  ASCII_BANNER_GENERATOR_SERVICE_ID,
} from "../../api/service/core/AsciiBannerGeneratorService.ts";
import type CLIConfig from "../../api/CLIConfig.ts";
export const BANNER_SERVICE_ID =
  "@flowscripter/dynamic-cli-framework/banner-service";

/**
 * Provides ascii banner functionality.
 */
export default class BannerServiceProvider implements ServiceProvider {
  readonly serviceId: string = BANNER_SERVICE_ID;
  readonly servicePriority: number;
  readonly #configurationServiceProvider:
    | ConfigurationServiceProvider
    | undefined;

  /**
   * Whether or not to print the banner, defaults to true.
   */
  printBanner = true;

  /**
   * Create an instance of the service provider with the specified details.
   *
   * @param servicePriority the priority of the service.
   * @param configurationServiceProvider an optional {@link ConfigurationServiceProvider} to make use of.
   * @param fontName an optional [FIGlet](http://www.figlet.org) font name to use for the banner, defaults to "standard".
   */
  public constructor(
    servicePriority: number,
    configurationServiceProvider?: ConfigurationServiceProvider,
    private readonly fontName: string = "standard",
  ) {
    this.servicePriority = servicePriority;
    this.#configurationServiceProvider = configurationServiceProvider;
  }

  public provide(_cliConfig: CLIConfig): Promise<ServiceInfo> {
    return Promise.resolve({
      commands: [
        new NoBannerCommand(this, this.servicePriority),
      ],
    });
  }

  async initService(context: Context): Promise<void> {
    if (this.printBanner === false) {
      return Promise.resolve();
    }

    const printerService = context.getServiceById(
      PRINTER_SERVICE_ID,
    ) as PrinterService;

    const asciiBannerGeneratorService = context.getServiceById(
      ASCII_BANNER_GENERATOR_SERVICE_ID,
    ) as AsciiBannerGeneratorService;

    const { cliConfig } = context;
    const bannerText = await asciiBannerGeneratorService.generate(
      cliConfig.name.toUpperCase(),
      this.fontName,
    );

    await printerService.info(printerService.blue(bannerText));
    if (cliConfig.description !== undefined) {
      await printerService.info(
        `  ${printerService.primary(cliConfig.description)}\n`,
      );
    }
    if (cliConfig.version.length > 0) {
      await printerService.info(
        `  ${printerService.secondary("version: " + cliConfig.version)}\n`,
      );
    }
    if (this.#configurationServiceProvider) {
      const configLocation = this.#configurationServiceProvider.configLocation;
      if (configLocation && configLocation.length > 0) {
        await printerService.info(
          `  ${
            printerService.secondary(
              "config: " + configLocation,
            )
          }\n`,
        );
      }
    }
    await printerService.info("\n");
  }
}
