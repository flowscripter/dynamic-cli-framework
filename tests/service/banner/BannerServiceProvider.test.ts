import { describe, expect, test } from "bun:test";
import { getCLIConfig } from "../../fixtures/CLIConfig.ts";
import BannerServiceProvider from "../../../src/service/banner/BannerServiceProvider.ts";
import DefaultPrinterService from "../../../src/service/printer/DefaultPrinterService.ts";
import { PRINTER_SERVICE_ID } from "../../../src/api/service/core/PrinterService.ts";
import DefaultContext from "../../../src/runtime/DefaultContext.ts";
import { ASCII_BANNER_GENERATOR_SERVICE_ID } from "../../../src/api/service/core/AsciiBannerGeneratorService.ts";
import DefaultAsciiBannerGeneratorService from "../../../src/service/asciiBannerGenerator/DefaultAsciiBannerGeneratorService.ts";
import TtyTerminal from "../../../src/service/printer/terminal/TtyTerminal.ts";
import StreamString from "../../fixtures/StreamString.ts";
import TtyStyler from "../../../src/service/printer/terminal/TtyStyler.ts";
import { getConfigurationServiceProvider } from "../../fixtures/ConfigurationServiceProvider.ts";

// FIGlet font is converted to a JSON string and embedded in a simple JSON file: `{ "font": "<figlet font definition>" }`
import smallFont from "../asciiBannerGenerator/small.flf.json" with {
  type: "json",
};

describe("BannerServiceProvider tests", () => {
  test("BannerServiceProvider provide works", async () => {
    const bannerServiceProvider = new BannerServiceProvider(100);
    const serviceInfo = await bannerServiceProvider.provide(getCLIConfig());
    expect(serviceInfo.commands.length).toEqual(1);
  });

  test("BannerServiceProvider initService works", async () => {
    const dummyStdout = new StreamString();
    const dummyStderr = new StreamString();
    const printer = new DefaultPrinterService(
      dummyStdout.writableStream,
      dummyStderr.writableStream,
      true,
      true,
      new TtyTerminal(dummyStderr.writeStream),
      new TtyStyler(3),
    );
    printer.colorEnabled = false;

    const asciiBannerGenerator = new DefaultAsciiBannerGeneratorService();
    const context = new DefaultContext(getCLIConfig());

    context.addServiceInstance(PRINTER_SERVICE_ID, printer);
    context.addServiceInstance(
      ASCII_BANNER_GENERATOR_SERVICE_ID,
      asciiBannerGenerator,
    );

    const bannerServiceProvider = new BannerServiceProvider(100);
    await bannerServiceProvider.initService(context);

    expect(bannerServiceProvider.printBanner).toBeTrue();

    expect(dummyStderr.getString()).toMatchSnapshot();
  });

  test("BannerServiceProvider with configuration service and custom font name works", async () => {
    const dummyStdout = new StreamString();
    const dummyStderr = new StreamString();
    const printer = new DefaultPrinterService(
      dummyStdout.writableStream,
      dummyStderr.writableStream,
      true,
      true,
      new TtyTerminal(dummyStderr.writeStream),
      new TtyStyler(3),
    );
    printer.colorEnabled = false;

    const asciiBannerGenerator = new DefaultAsciiBannerGeneratorService();
    const context = new DefaultContext(getCLIConfig("mpeg-sdl-tool"));

    context.addServiceInstance(PRINTER_SERVICE_ID, printer);
    context.addServiceInstance(
      ASCII_BANNER_GENERATOR_SERVICE_ID,
      asciiBannerGenerator,
    );

    asciiBannerGenerator.registerFont("small", smallFont.font);

    const configurationServiceProvider = getConfigurationServiceProvider(
      100,
      new Map(),
    );
    configurationServiceProvider.configLocation = "config.yaml";

    const bannerServiceProvider = new BannerServiceProvider(
      100,
      configurationServiceProvider,
      "small",
    );
    await bannerServiceProvider.initService(context);
    await printer.error("some text\n");
    expect(bannerServiceProvider.printBanner).toBeTrue();
    expect(dummyStdout.getString()).toMatchSnapshot();
  });
});
