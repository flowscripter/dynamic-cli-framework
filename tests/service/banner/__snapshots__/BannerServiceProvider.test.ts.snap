// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`BannerServiceProvider tests BannerServiceProvider initService works 1`] = `
"
  _____ ___   ___  
 |  ___/ _ \\ / _ \\ 
 | |_ | | | | | | |
 |  _|| |_| | |_| |
 |_|   \\___/ \\___/ 
  bar
  version: foobar

"
`;

exports[`BannerServiceProvider tests BannerServiceProvider with configuration service and custom font name works 1`] = `""`;
